File: Opsec/Opsec.go
// test794c : USAG-Lib opsec

package Opsec

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/subtle"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"hash/crc32"
	"io"
	"math/bits"
	"runtime"
	"sync"
	"sync/atomic"

	Bencrypt "github.com/k-atusa/USAG-Lib/v2/Bencrypt"
)

// ========== Helper Functions ==========
var SCLEAR_BACK = func(b []byte) { clear(b) }

func sclear(data []byte) { SCLEAR_BACK(data); runtime.KeepAlive(data) }

func Crc32(data []byte) string {
	checksum := crc32.ChecksumIEEE(data)
	buf := make([]byte, 4)
	binary.LittleEndian.PutUint32(buf, checksum)
	return hex.EncodeToString(buf)
}

func PadLen(size int64) int64 {
	if size <= 0 {
		return 0
	}

	// 1. 0-16k: 4k*N
	if size <= 16384 {
		remainder := size % 4096
		if remainder == 0 {
			return 0
		}
		return 4096 - remainder
	}

	// get sup bit position
	bitLen := bits.Len64(uint64(size))
	var k int
	if bitLen <= 24 { // 16k-16m: K=2
		k = 2
	} else if bitLen <= 29 { // 16m-512m: K=3
		k = 3
	} else if bitLen <= 33 { // 512m-8g: K=4
		k = 4
	} else { // 8g+: K=5
		k = 5
	}

	// mask and ceiling
	shift := bitLen - k
	mask := (int64(1) << shift) - 1
	if size&mask == 0 { // on border size is not padded
		return 0
	}

	// return actual padding length
	aftersize := ((size >> shift) + 1) << shift
	return aftersize - size
}

func PadFile(f io.Writer, size int64) error {
	if size <= 0 {
		return nil
	}
	chunkSize := 128 * 1024         // 128 KiB
	rotationSize := 8 * 1024 * 1024 // 8 MiB

	// shared pools
	zeroBuf := make([]byte, chunkSize)
	bufPool := sync.Pool{
		New: func() any {
			return make([]byte, chunkSize)
		},
	}
	numWorkers := max(runtime.NumCPU(), 2)
	var stopFlag atomic.Bool
	stopFlag.Store(false)
	results := make(chan []byte, numWorkers*4)

	// writer
	var wg sync.WaitGroup
	errCh := make(chan error, 1)
	go func() {
		var writeErr error
		for buf := range results {
			if writeErr == nil {
				if _, err := f.Write(buf); err != nil {
					stopFlag.Store(true)
					writeErr = err
				}
			}
			bufPool.Put(buf) // return used buffer
		}
		errCh <- writeErr
	}()

	// Random generator
	blockerChan := make(chan bool, numWorkers)
	remainingSize := size
	for remainingSize > 0 {
		currentSegSize := min(remainingSize, int64(rotationSize))
		if stopFlag.Load() {
			break
		}

		wg.Add(1)
		blockerChan <- true
		go func(segSize int64) {
			defer wg.Done()
			defer func() { <-blockerChan }()
			key := Bencrypt.Random(48)
			block, _ := aes.NewCipher(key[16:48]) // AES-CTR CSPRNG
			stream := cipher.NewCTR(block, key[0:16])

			for i := int64(0); i < segSize; i += int64(chunkSize) {
				if stopFlag.Load() {
					return
				}
				currChunkSize := min(segSize-i, int64(chunkSize))

				fullBuf := bufPool.Get().([]byte)
				buf := fullBuf[:cap(fullBuf)][:currChunkSize]
				stream.XORKeyStream(buf, zeroBuf[:currChunkSize])
				results <- buf
			}
		}(currentSegSize)
		remainingSize -= currentSegSize
	}

	wg.Wait()
	close(blockerChan)
	close(results)
	return <-errCh
}

func EncodeInt(data uint64, size int) []byte {
	buf := make([]byte, size)
	switch size {
	case 1:
		buf[0] = byte(data)
	case 2:
		binary.LittleEndian.PutUint16(buf, uint16(data))
	case 4:
		binary.LittleEndian.PutUint32(buf, uint32(data))
	case 8:
		binary.LittleEndian.PutUint64(buf, data)
	}
	return buf
}

func DecodeInt(data []byte) uint64 {
	l := len(data)
	switch l {
	case 1:
		return uint64(data[0])
	case 2:
		return uint64(binary.LittleEndian.Uint16(data))
	case 4:
		return uint64(binary.LittleEndian.Uint32(data))
	case 8:
		return binary.LittleEndian.Uint64(data)
	default:
		return 0
	}
}

// Config Encoding, keysize max 127, datasize max 65535
func EncodeCfg(data map[string][]byte) ([]byte, error) {
	var buf bytes.Buffer
	for key, val := range data {
		keyBytes := []byte(key)
		keyLen := len(keyBytes)
		dataLen := len(val)
		if keyLen > 127 {
			return nil, errors.New("key length too long")
		}
		if dataLen > 65535 {
			return nil, errors.New("data size too big")
		}

		if dataLen > 255 {
			buf.WriteByte(byte(keyLen + 128))
			buf.Write(keyBytes)
			buf.Write(EncodeInt(uint64(dataLen), 2))
		} else {
			buf.WriteByte(byte(keyLen))
			buf.Write(keyBytes)
			buf.WriteByte(byte(dataLen))
		}
		buf.Write(val)
	}
	return buf.Bytes(), nil
}

// Config Decoding, [keyLen 1B][key][dataLen 1B/2B][data]
func DecodeCfg(data []byte) map[string][]byte {
	result := make(map[string][]byte)
	totalLen := len(data)
	offset := 0
	for offset < totalLen {
		// Get Key
		keyLen := int(data[offset])
		offset++
		isLongData := false
		if keyLen > 127 {
			keyLen -= 128
			isLongData = true
		}
		key := string(data[offset : offset+keyLen])
		offset += keyLen

		// Get Data
		var dataLen int
		if isLongData {
			dataLen = int(DecodeInt(data[offset : offset+2]))
			offset += 2
		} else {
			dataLen = int(data[offset])
			offset++
		}
		result[key] = data[offset : offset+dataLen]
		offset += dataLen
	}
	return result
}

// Opsec header handler
type Opsec struct {
	// Outer Layer
	Msg         string // non-secured message
	MsgInfo     []byte // additional info (for RSA-mode)
	headAlgo    string // header algorithm
	salt        []byte // salt
	pwHash      []byte // pw hash
	encHeadData []byte // encrypted header data

	// Inner Layer
	Smsg     string // secured message
	SmsgInfo []byte // private additional info (timestamp, ID, etc.)
	sign     []byte // signature

	BodyAlgo string // body algorithm
	BodyKey  []byte // body key
	BodySize int64  // full body size, flag for bodyKey generation
	BodyInfo []byte // additional info for body (packing info, etc.)

	SaltLen int
}

func (o *Opsec) Clear() {
	sclear(o.salt)
	sclear(o.pwHash)
	sclear(o.encHeadData)
	sclear(o.MsgInfo)
	sclear(o.SmsgInfo)
	sclear(o.sign)
	sclear(o.BodyKey)
	sclear(o.BodyInfo)
	o.Init()
}

// set initial values
func (o *Opsec) Init() {
	o.Msg = ""
	o.MsgInfo = []byte{}
	o.headAlgo = ""
	o.salt = []byte{}
	o.pwHash = []byte{}
	o.encHeadData = []byte{}

	o.Smsg = ""
	o.SmsgInfo = []byte{}
	o.sign = []byte{}

	o.BodyAlgo = ""
	o.BodyKey = []byte{}
	o.BodySize = -1
	o.BodyInfo = []byte{}

	o.SaltLen = 32
}

// Read opsec header from stream, set cut to 0 to read all
func (o *Opsec) Read(r io.Reader, cut int) ([]byte, error) {
	c := 0
	buf4 := make([]byte, 4)
	buf2 := make([]byte, 2)
	buf124 := make([]byte, 124)
	for {
		_, err := io.ReadFull(r, buf4)
		if err != nil { // EOF or Error
			return nil, nil
		}
		c += 4

		if string(buf4) == "YAS2" { // magic number
			_, err := io.ReadFull(r, buf2)
			if err != nil {
				return nil, err
			}
			size := int(DecodeInt(buf2))
			if size == 65535 {
				_, err := io.ReadFull(r, buf2)
				if err != nil {
					return nil, err
				}
				size += int(DecodeInt(buf2))
			}

			// Read payload
			payload := make([]byte, size)
			_, err = io.ReadFull(r, payload)
			if err != nil {
				return nil, err
			}
			return payload, nil

		} else { // skip
			_, err := io.ReadFull(r, buf124)
			if err != nil {
				return nil, nil
			}
			c += 124
		}
		if cut > 0 && c > cut {
			return nil, nil
		}
	}
}

// Write opsec header to stream
func (o *Opsec) Write(w io.Writer, head []byte) error {
	if _, err := w.Write([]byte("YAS2")); err != nil {
		return err
	}
	size := len(head)
	if size < 65535 {
		if _, err := w.Write(EncodeInt(uint64(size), 2)); err != nil {
			return err
		}
	} else if size <= 65535*2 {
		if _, err := w.Write(EncodeInt(65535, 2)); err != nil {
			return err
		}
		if _, err := w.Write(EncodeInt(uint64(size-65535), 2)); err != nil {
			return err
		}
	} else {
		return errors.New("data size too big")
	}
	if _, err := w.Write(head); err != nil {
		return err
	}
	return nil
}

func (o *Opsec) wrapEncHead() ([]byte, error) {
	cfg := make(map[string][]byte)
	if o.Smsg != "" {
		cfg["smsg"] = []byte(o.Smsg)
	}
	if len(o.SmsgInfo) > 0 {
		cfg["sinf"] = o.SmsgInfo
	}
	if len(o.sign) > 0 {
		cfg["sgn"] = o.sign
	}
	if o.BodyAlgo != "" {
		cfg["bal"] = []byte(o.BodyAlgo)
	}
	if len(o.BodyKey) > 0 {
		cfg["bkey"] = o.BodyKey
	}
	if o.BodySize >= 0 {
		var szBytes []byte
		if o.BodySize < 65536 {
			szBytes = EncodeInt(uint64(o.BodySize), 2)
		} else if o.BodySize < 4294967296 {
			szBytes = EncodeInt(uint64(o.BodySize), 4)
		} else {
			szBytes = EncodeInt(uint64(o.BodySize), 8)
		}
		cfg["bsz"] = szBytes
	}
	if len(o.BodyInfo) > 0 {
		cfg["binf"] = o.BodyInfo
	}
	return EncodeCfg(cfg)
}

func (o *Opsec) unwrapEncHead(data []byte) {
	cfg := DecodeCfg(data)
	if v, ok := cfg["smsg"]; ok {
		o.Smsg = string(v)
	}
	if v, ok := cfg["sinf"]; ok {
		o.SmsgInfo = v
	}
	if v, ok := cfg["sgn"]; ok {
		o.sign = v
	}
	if v, ok := cfg["bal"]; ok {
		o.BodyAlgo = string(v)
	}
	if v, ok := cfg["bkey"]; ok {
		o.BodyKey = v
	}
	if v, ok := cfg["bsz"]; ok {
		o.BodySize = int64(DecodeInt(v))
	}
	if v, ok := cfg["binf"]; ok {
		o.BodyInfo = v
	}
}

// Encrypt with password
func (o *Opsec) Encpw(method string, pw []byte, kf []byte) ([]byte, error) {
	// generate random parameters
	if o.SaltLen <= 0 {
		o.SaltLen = 32
	}
	o.headAlgo = method
	o.salt = Bencrypt.Random(o.SaltLen)
	if o.BodySize >= 0 {
		o.BodyKey = Bencrypt.Random(32)
	}

	// Combine pw + kf
	combinedPw := make([]byte, 0, len(pw)+len(kf))
	combinedPw = append(combinedPw, pw...)
	combinedPw = append(combinedPw, kf...)
	defer sclear(combinedPw)

	// KDF via HashMaster
	hm := new(Bencrypt.HashMaster)
	if err := hm.Init(method, 0, 0); err != nil {
		return nil, err
	}
	pwHash, hkey, err := hm.KDF(combinedPw, o.salt)
	defer sclear(hkey)
	if err != nil {
		return nil, err
	}
	o.pwHash = pwHash

	// Encrypt header
	headData, err := o.wrapEncHead()
	defer sclear(headData)
	if err != nil {
		return nil, err
	}
	sm := new(Bencrypt.SymMaster)
	if err := sm.Init("gcm1", hkey); err != nil {
		return nil, err
	}
	o.encHeadData, err = sm.EnBin(headData)
	if err != nil {
		return nil, err
	}

	// Wrap outer header
	cfg := make(map[string][]byte)
	if o.Msg != "" {
		cfg["msg"] = []byte(o.Msg)
	}
	if len(o.MsgInfo) > 0 {
		cfg["minf"] = o.MsgInfo
	}
	cfg["hal"] = []byte(o.headAlgo)
	cfg["salt"] = o.salt
	cfg["pwh"] = o.pwHash
	cfg["ehd"] = o.encHeadData
	return EncodeCfg(cfg)
}

// Encrypt with public key, sign if private key is not nil
func (o *Opsec) Encpub(method string, peerPub []byte, myPri []byte) ([]byte, error) {
	o.headAlgo = method
	if o.BodySize >= 0 {
		o.BodyKey = Bencrypt.Random(32)
	}

	// Sign if private key is provided
	if myPri != nil {
		amSign := new(Bencrypt.AsymMaster)
		if err := amSign.Init(method); err != nil {
			return nil, err
		}
		if err := amSign.Loadkey(nil, myPri); err != nil {
			return nil, err
		}

		signTgt := make([]byte, 0)
		signTgt = append(signTgt, []byte(method)...)
		signTgt = append(signTgt, peerPub...)
		signTgt = append(signTgt, []byte(o.Smsg)...)
		signTgt = append(signTgt, o.SmsgInfo...)
		defer sclear(signTgt)

		var err error
		o.sign, err = amSign.Sign(signTgt)
		if err != nil {
			return nil, err
		}
	}

	// Encrypt header
	amEncrypt := new(Bencrypt.AsymMaster)
	if err := amEncrypt.Init(method); err != nil {
		return nil, err
	}
	if err := amEncrypt.Loadkey(peerPub, nil); err != nil {
		return nil, err
	}

	headData, err := o.wrapEncHead()
	defer sclear(headData)
	if err != nil {
		return nil, err
	}

	o.encHeadData, err = amEncrypt.Encrypt(headData)
	if err != nil {
		return nil, err
	}

	// Wrap outer header
	cfg := make(map[string][]byte)
	if o.Msg != "" {
		cfg["msg"] = []byte(o.Msg)
	}
	if len(o.MsgInfo) > 0 {
		cfg["minf"] = o.MsgInfo
	}
	cfg["hal"] = []byte(o.headAlgo)
	cfg["ehd"] = o.encHeadData
	return EncodeCfg(cfg)
}

// Load outer layer of header
func (o *Opsec) View(data []byte) {
	o.Init()
	cfg := DecodeCfg(data)
	if v, ok := cfg["msg"]; ok {
		o.Msg = string(v)
	}
	if v, ok := cfg["minf"]; ok {
		o.MsgInfo = v
	}
	if v, ok := cfg["hal"]; ok {
		o.headAlgo = string(v)
	}
	if v, ok := cfg["salt"]; ok {
		o.salt = v
	}
	if v, ok := cfg["pwh"]; ok {
		o.pwHash = v
	}
	if v, ok := cfg["ehd"]; ok {
		o.encHeadData = v
	}
}

// Decrypt with password
func (o *Opsec) Decpw(pw []byte, kf []byte) error {
	if o.headAlgo == "" {
		return errors.New("call View() first")
	}

	combinedPw := make([]byte, 0, len(pw)+len(kf))
	combinedPw = append(combinedPw, pw...)
	combinedPw = append(combinedPw, kf...)
	defer sclear(combinedPw)

	// KDF via HashMaster
	hm := new(Bencrypt.HashMaster)
	if err := hm.Init(o.headAlgo, 0, 0); err != nil {
		return err
	}
	calcHash, hkey, err := hm.KDF(combinedPw, o.salt)
	defer sclear(hkey)
	if err != nil {
		return err
	}

	// Constant time comparison
	if subtle.ConstantTimeCompare(calcHash, o.pwHash) != 1 {
		return errors.New("incorrect password")
	}

	// Decrypt header
	sm := new(Bencrypt.SymMaster)
	if err := sm.Init("gcm1", hkey); err != nil {
		return err
	}
	decryptedHead, err := sm.DeBin(o.encHeadData)
	if err != nil {
		return fmt.Errorf("AES decryption failed: %w", err)
	}
	o.unwrapEncHead(decryptedHead)
	return nil
}

// Decrypt with private key, verify if public key is not nil
func (o *Opsec) Decpub(myPri []byte, myPub []byte, peerPub []byte) error {
	if o.headAlgo == "" {
		return errors.New("call View() first")
	}

	// Init AsymMaster
	am := new(Bencrypt.AsymMaster)
	if err := am.Init(o.headAlgo); err != nil {
		return err
	}
	if err := am.Loadkey(nil, myPri); err != nil {
		return err
	}

	// Decrypt header
	var decryptedHead []byte
	var err error

	decryptedHead, err = am.Decrypt(o.encHeadData)
	if err != nil {
		return fmt.Errorf("Asymmetric decryption failed: %w", err)
	}
	o.unwrapEncHead(decryptedHead)

	// Verify if public keys are provided
	if myPub == nil && peerPub == nil {
		return nil
	}
	if myPub == nil || peerPub == nil {
		if len(o.sign) > 0 {
			return errors.New("both myPub and peerPub should be provided to verify sign")
		}
		return nil
	}

	amVerify := new(Bencrypt.AsymMaster)
	if err := amVerify.Init(o.headAlgo); err != nil {
		return err
	}
	if err := amVerify.Loadkey(peerPub, nil); err != nil {
		return err
	}

	signTgt := make([]byte, 0)
	signTgt = append(signTgt, []byte(o.headAlgo)...)
	signTgt = append(signTgt, myPub...)
	signTgt = append(signTgt, []byte(o.Smsg)...)
	signTgt = append(signTgt, o.SmsgInfo...)

	if !amVerify.Verify(signTgt, o.sign) {
		return errors.New("signature verification failed")
	}
	return nil
}

File: Opsec/Opsec.py
# test794a : USAG-Lib opsec
from typing import Dict, Union

import io
import zlib
import Bencrypt

def Crc32(data: bytes) -> str:
    return zlib.crc32(data).to_bytes(4, 'little').hex()

def PadLen(size: int) -> int:
    if size <= 0:
        return 0

    # 1. 0-16k: 4k*N
    if size <= 16384:
        remainder = size % 4096
        if remainder == 0:
            return 0
        return 4096 - remainder

    # get sup bit position
    bitLen = size.bit_length()
    if bitLen <= 24: # 16k-16m: K=2
        k = 2
    elif bitLen <= 29: # 16m-512m: K=3
        k = 3
    elif bitLen <= 33: # 512m-8g: K=4
        k = 4
    else: # 8g+: K=5
        k = 5

    # mask and ceiling
    shift = bitLen - k
    mask = (1 << shift) - 1
    if size & mask == 0: # on border size is not padded
        return 0

    # return actual padding length
    aftersize = ((size >> shift) + 1) << shift
    return aftersize - size

def PadFile(f: io.IOBase, size: int):
    for i in range(0, size // 1048576):
        f.write( Bencrypt.Random(1048576) )
    if size % 1048576 != 0:
        f.write( Bencrypt.Random(size % 1048576) )

def EncodeInt(data: int, size: int, signed: bool) -> bytes:
    return data.to_bytes(size, 'little', signed=signed)

def DecodeInt(data: bytes, signed: bool) -> int:
    return int.from_bytes(data, 'little', signed=signed)

def EncodeCfg(data: Dict[str, bytes]) -> bytes: # keysize max 127, datasize max 65535
    result = bytearray()
    for key, data in data.items():
        keyBytes = key.encode('utf-8')
        keyLen = len(keyBytes)
        dataLen = len(data)
        if keyLen > 127:
            raise ValueError(f"Key length too long: {keyLen}")
        if dataLen > 65535:
            raise ValueError(f"Data size too big: {dataLen}")
        if dataLen > 255: # dataLen > 255, datasize is 2B
            encodedKeyLen = keyLen + 128
            result.append(encodedKeyLen)
            result.extend(keyBytes)
            result.extend(dataLen.to_bytes(2, 'little'))
        else: # dataLen <= 255, datasize is 1B
            result.append(keyLen)
            result.extend(keyBytes)
            result.append(dataLen)
        result.extend(data)
    return bytes(result)

def DecodeCfg(data: bytes) -> Dict[str, bytes]: # format: [keyLen 1B][key][dataLen 1B/2B][data]
    result: Dict[str, bytes] = {}
    offset = 0
    totalLen = len(data)
    while offset < totalLen:
        keyLen = data[offset] # read keyLen
        isLongData = False
        offset += 1
        if keyLen > 127:
            keyLen -= 128
            isLongData = True
        keyBytes = data[offset : offset + keyLen] # read key
        key = keyBytes.decode('utf-8')
        offset += keyLen
        if isLongData: # dataLen is 2B
            dataLen = int.from_bytes(data[offset : offset + 2], 'little')
            offset += 2
        else: # dataLen is 1B
            dataLen = data[offset]
            offset += 1
        result[key] = data[offset : offset + dataLen] # read data
        offset += dataLen
    return result

# Opsec header handler
class Opsec:
    def __init__(self):
        self.Init()
        self.SaltLen = 32

    def __del__(self):
        self.Clear()
        if hasattr(self, 'BodyKey'):
            del self.BodyKey

    # set initial values
    def Init(self):
        self._headAlgo: str = ""
        self.Msg: str = "" # public message
        self.MsgInfo: bytes = b"" # additional info

        self._salt: bytes = b""
        self._encHeadData: bytes = b""

        self.Smsg: str = "" # private message
        self.SmsgInfo: bytes = b"" # private additional info (timestamp, ID, etc.)
        self._sign: bytes = b""

        self.BodyAlgo: str = "" # body encryption algorithm
        self.BodyKey: bytes = b"" # body encryption key
        self.BodySize: int = -1 # body size (-1 if not used)
        self.BodyInfo: bytes = b"" # additional info for body (packing info, etc.)

    # clear all (not practical, but to match with other codes)
    def Clear(self):
        self._salt = b""
        self._encHeadData = b""
        self.MsgInfo = b""
        self.SmsgInfo = b""
        self._sign = b""
        self.BodyKey = b""
        self.BodyInfo = b""
        self.Init()

    def Read(self, ins: io.IOBase, cut: int = 65535) -> bytes: # set cut to 0 to read all
        c = 0
        while True:
            data = ins.read(4)
            c += 4
            if data == b"":
                return b""
            elif data == b"YAS2":
                size = DecodeInt(ins.read(2), False)
                if size == 65535:
                    size += DecodeInt(ins.read(2), False)
                return ins.read(size)
            else:
                ins.read(124)
                c += 124
            if cut > 0 and c > cut:
                return b""

    def Write(self, outs: io.IOBase, head: bytes):
        outs.write(b"YAS2")
        size = len(head)
        if size < 65535:
            outs.write(EncodeInt(size, 2, False))
        elif size <= 65535 * 2:
            outs.write(EncodeInt(65535, 2, False))
            outs.write(EncodeInt(size - 65535, 2, False))
        else:
            raise ValueError(f"Header too big: {size}")
        outs.write(head)

    def _wrapEncHead(self) -> bytes:
        cfg: Dict[str, bytes] = {}
        if self.Smsg != "":
            cfg["smsg"] = self.Smsg.encode('utf-8')
        if self.SmsgInfo != b"":
            cfg["sinf"] = self.SmsgInfo
        if self._sign != b"":
            cfg["sgn"] = self._sign
        if self.BodyAlgo != "":
            cfg["bal"] = self.BodyAlgo.encode('utf-8')
        if self.BodyKey != b"":
            cfg["bkey"] = self.BodyKey
        if self.BodySize >= 0:
            if self.BodySize < 65536:
                cfg["bsz"] = EncodeInt(self.BodySize, 2, False)
            elif self.BodySize < 4294967296:
                cfg["bsz"] = EncodeInt(self.BodySize, 4, False)
            else:
                cfg["bsz"] = EncodeInt(self.BodySize, 8, False)
        if self.BodyInfo != b"":
            cfg["binf"] = self.BodyInfo
        return EncodeCfg(cfg)
    
    def _unwrapEncHead(self, data: bytes):
        cfg = DecodeCfg(data)
        if "smsg" in cfg:
            self.Smsg = cfg["smsg"].decode('utf-8')
        if "sinf" in cfg:
            self.SmsgInfo = cfg["sinf"]
        if "sgn" in cfg:
            self._sign = cfg["sgn"]
        if "bal" in cfg:
            self.BodyAlgo = cfg["bal"].decode('utf-8')
        if "bkey" in cfg:
            self.BodyKey = cfg["bkey"]
        if "bsz" in cfg:
            self.BodySize = DecodeInt(cfg["bsz"], False)
        if "binf" in cfg:
            self.BodyInfo = cfg["binf"]

    def Encpw(self, method: str, pw: bytes, kf: bytes = b"") -> bytes:
        # generate random parameters
        self._headAlgo = method
        self._salt = Bencrypt.Random(self.SaltLen)
        if self.BodySize >= 0:
            self.BodyKey = Bencrypt.Random(32)

        # get header key, encrypt header
        hm = Bencrypt.HashMaster(method)
        _, hkey = hm.KDF(pw + kf, self._salt)
        headData = self._wrapEncHead()
        sm = Bencrypt.SymMaster("gcm1", hkey)
        del hkey
        self._encHeadData = sm.EnBin(headData)
        del headData

        # warp header
        cfg: Dict[str, bytes] = {}
        if self.Msg != "":
            cfg["msg"] = self.Msg.encode('utf-8')
        if self.MsgInfo != b"":
            cfg["minf"] = self.MsgInfo
        cfg["hal"] = self._headAlgo.encode('utf-8')
        cfg["salt"] = self._salt
        cfg["ehd"] = self._encHeadData
        return EncodeCfg(cfg)
    
    def Encpub(self, method: str, peerPub: bytes, myPri: Union[bytes, None] = None) -> bytes: # sign if private is not None
        # generate random parameters
        self._headAlgo = method
        if self.BodySize >= 0:
            self.BodyKey = Bencrypt.Random(32)
        
        # sign with private key if provided
        if myPri != None:
            am = Bencrypt.AsymMaster(method)
            am.Loadkey(None, myPri)
            # sign to [hal][peerPub][smsg][sinf] with 0-byte suffix for each field
            signTgt = method.encode('utf-8') + b'\x00' + peerPub + b'\x00' + self.Smsg.encode('utf-8') + b'\x00' + self.SmsgInfo + b'\x00'
            self._sign = am.Sign(signTgt)

        # encrypt header
        am = Bencrypt.AsymMaster(method) # Bencrypt will check if method is valid
        am.Loadkey(peerPub, None)
        headData = self._wrapEncHead()
        self._encHeadData = am.Encrypt(headData)
        del headData

        # warp header
        cfg: Dict[str, bytes] = {}
        if self.Msg != "":
            cfg["msg"] = self.Msg.encode('utf-8')
        if self.MsgInfo != b"":
            cfg["minf"] = self.MsgInfo
        cfg["hal"] = self._headAlgo.encode('utf-8')
        cfg["ehd"] = self._encHeadData
        return EncodeCfg(cfg)
        
    def View(self, data: bytes):
        self.Init()
        cfg = DecodeCfg(data)
        if "msg" in cfg:
            self.Msg = cfg["msg"].decode('utf-8')
        if "minf" in cfg:
            self.MsgInfo = cfg["minf"]
        if "hal" in cfg:
            self._headAlgo = cfg["hal"].decode('utf-8')
        if "salt" in cfg:
            self._salt = cfg["salt"]
        if "ehd" in cfg:
            self._encHeadData = cfg["ehd"]

    def Decpw(self, pw: bytes, kf: bytes = b""):
        # check parameters, get header key
        if self._headAlgo == "":
            raise ValueError("Opsec not initialized or invalid data")
        hm = Bencrypt.HashMaster(self._headAlgo)
        _, hkey = hm.KDF(pw + kf, self._salt)

        # decrypt header (verification by SymMaster)
        sm = Bencrypt.SymMaster("gcm1", hkey)
        del hkey
        headData = sm.DeBin(self._encHeadData)
        self._unwrapEncHead(headData)
        del headData

    def Decpub(self, myPri: bytes, myPub: Union[bytes, None] = None, peerPub: Union[bytes, None] = None): # verify sign if public is not None
        # check parameters, decrypt header
        if self._headAlgo == "":
            raise ValueError("Opsec not initialized or invalid data")
        am = Bencrypt.AsymMaster(self._headAlgo)
        am.Loadkey(None, myPri)
        headData = am.Decrypt(self._encHeadData)
        self._unwrapEncHead(headData)
        del headData

        # verify sign
        if myPub == None and peerPub == None:
            return
        if myPub == None or peerPub == None:
            raise ValueError("Both myPub and peerPub should be provided to verify sign")
        am = Bencrypt.AsymMaster(self._headAlgo)
        am.Loadkey(peerPub, None)
        # verify sign [hal][myPub][smsg][sinf] with 0-byte suffix for each field
        signTgt = self._headAlgo.encode('utf-8') + b'\x00' + myPub + b'\x00' + self.Smsg.encode('utf-8') + b'\x00' + self.SmsgInfo + b'\x00'
        if not am.Verify(signTgt, self._sign):
            raise ValueError("Sign verification failed")